byterun/array.c

```diff
@@ -69,10 +69,31 @@ CAMLprim value caml_array_get_float(value array, value index)
 
 CAMLprim value caml_array_get(value array, value index)
 {
+#ifdef FLAT_FLOAT_ARRAY
   if (Tag_val(array) == Double_array_tag)
     return caml_array_get_float(array, index);
-  else
-    return caml_array_get_addr(array, index);
+#endif
+  CAMLassert (Tag_val(array) != Double_array_tag);
+  return caml_array_get_addr(array, index);
+}
+
+CAMLprim value caml_floatarray_get(value array, value index)
```

Comment:
As said in another note, I believe this should become a %-primitive and be handled directly by the compiler.